### PR TITLE
Move exponential backoff into a function

### PR DESCRIPTION
**Pull request by yuvipanda**
Also use the 'Full Jitter' jitter algorithm from
https://www.awsarchitectureblog.com/2015/03/backoff.html

### CODE REVIEW

File: jupyterhub/objects.py [hunk 1]
```diff
@@ -115,12 +115,13 @@ def bind_url(self):
             return self.url.replace(self._connect_ip, self.ip or '*', 1)
         return self.url
 
+    @gen.coroutine
     def wait_up(self, timeout=10, http=False):
         """Wait for this server to come up"""
         if http:
-            return wait_for_http_server(self.url, timeout=timeout)
+            return (yield wait_for_http_server(self.url, timeout=timeout))
```
Comment (minrk) [inline jupyterhub/objects.py]: Why this change? It adds an additional async layer that shouldn't be needed if we return the underlying result directly.

File: jupyterhub/spawner.py [hunk 2]
```diff
@@ -652,21 +652,19 @@ def poll_and_notify(self):
...
+        def _wait_for_death():
             status = yield self.poll()
-            if status is not None:
-                break
-            else:
-                yield gen.sleep(dt)
-            dt = min(dt * DT_SCALE, DT_MAX, timeout - (loop.time() - tic))
+            return status is not None
+
+        yield exponential_backoff(
```
Comment (minrk) [inline jupyterhub/spawner.py]: I'd like to keep `start_wait=self.death_interval` so subclasses can slow this down if needed.

File: jupyterhub/utils.py [hunk 3]
```diff
@@ -80,13 +114,13 @@ def wait_for_http_server(url, timeout=10):
...
-    loop = ioloop.IOLoop.current()
-    tic = loop.time()
     client = AsyncHTTPClient()
-    dt = DT_MIN
-    while dt > 0:
+    @gen.coroutine
+    def is_reachable():
         try:
             r = yield client.fetch(url, follow_redirects=False)
+            print('returning r!')
```
Comment (minrk) [inline jupyterhub/utils.py]: print statement

File: jupyterhub/utils.py [hunk 4]
```diff
@@ -48,29 +49,62 @@ def can_connect(ip, port):
...
+            return ret
+        else:
+            yield gen.sleep(dt)
+        # Add some random jitter to improve performance
+        # This makes sure that we don't overload any single iteration
+        # of the tornado loop with too many things
+        # See https://www.awsarchitectureblog.com/2015/03/backoff.html
+        # for a good example of why and how this helps. We're using their
+        # full Jitter implementation equivalent.
+        dt = min(max_wait, random.uniform(0, dt * scale_factor))
```
Comment (minrk) [inline jupyterhub/utils.py]: `loop.time() - start_tic` should also be added here as an upper limit. We don't want the last call to take 5 seconds if we've already waited 8 out of 10. See the implementation in master:

```python
while dt > 0:
    ...
    dt = min(max_wait, random(...), timeout - (loop.time() - start_tic))
```

which will turn negative and exit the loop when `loop.time() - start_tic` exceeds timeout.

Comment (minrk) [inline jupyterhub/utils.py]: Another point: probably want to introduce jitter on the first `dt` before `gen.sleep`, right? Rather than only after the second iteration? I think the dt calculation also probably makes more sense before gen.sleep. Maybe something like this:

```python
if ret:
    return ret
deadline = timeout - (loop.time() - start_tic)
if deadline < 0:
    # timeout exceeded
    break
dt = min(max_wait, deadline, random.uniform(0, dt * scale_factor))
yield gen.sleep(dt)
```

File: jupyterhub/spawner.py [hunk 5]
```diff
@@ -666,21 +666,24 @@ def poll_and_notify(self):
                 self.log.exception("Unhandled error in poll callback for %s", self)
         return status
 
-    death_interval = Float(DT_MIN)
-
+    death_interval = Float(0.1)
```
Comment (willingc) [inline jupyterhub/spawner.py]: Perhaps add a brief comment on what the death_interval is. Assuming 0.1s and is used when adding random jitter.

File: jupyterhub/spawner.py [hunk 6]
```diff
@@ -666,21 +666,24 @@ def poll_and_notify(self):
...
+        @gen.coroutine
+        def _wait_for_death():
             status = yield self.poll()
-            if status is not None:
-                break
-            else:
-                yield gen.sleep(dt)
-            dt = min(dt * DT_SCALE, DT_MAX, timeout - (loop.time() - tic))
+            return status is not None
+        try:
```
Comment (willingc) [inline jupyterhub/spawner.py]: Add LF between 676 and 677.

File: jupyterhub/utils.py [hunk 7]
```diff
@@ -48,29 +49,69 @@ def can_connect(ip, port):
...
+    deadline = loop.time() + timeout
+    # add some jitter to the deadline itself, so that we don't
+    # re-align a bunch of timing out calls once the deadline is reached.
+    if timeout_tolerance:
+        tol = timeout_tolerance * timeout
+        deadline = random.uniform(deadline - tol, deadline + tol)
+    scale = 1
+    while True:
+        ret = yield gen.maybe_future(pass_func(*args, **kwargs))
+        # Truthy!
```
Comment (willingc) [inline jupyterhub/utils.py]: Perhaps:
`# Return pass_func value when truthy`

File: jupyterhub/utils.py [hunk 8]
```diff
@@ -48,29 +49,69 @@ def can_connect(ip, port):
...
+        pass_func,
+        fail_message,
+        start_wait=0.2,
+        scale_factor=2,
+        max_wait=5,
+        timeout=10,
+        timeout_tolerance=0.1,
+        *args, **kwargs):
+    """
+    Exponentially backoff until pass_func is true.
```
Comment (willingc) [inline jupyterhub/utils.py]: Going to edit the docstring

**Comment by willingc** [comment]
@yuvipanda Thanks!

**Comment by yuvipanda** [comment]
This looks good to me! tvym for the cleanups, @minrk and @willingc!

**Comment by willingc** [comment]
And tyvm for the cherrypick. 🍨 :tea: 💤 for you
